// @keystone-next/fields/types/timestamp/views/dist/fields.esm.js
import { useState, Fragment } from 'react';
import { jsx } from '@keystone-ui/core';
import { FieldContainer, FieldLabel, TextInput } from '@keystone-ui/fields';
import { CellLink, CellContainer } from '@keystone-next/admin-ui/components';
import { parseISO } from 'date-fns';

/* @jsx jsx */

function formatOutput(value) {
  if (!value) return '';
  const date = new Date(value);
  return `${date.toLocaleDateString()} ${date.toLocaleTimeString()}`;
}

const Field = ({
  field,
  value,
  onChange,
  forceValidation,
  autoFocus
}) => {
  const [touched, setTouched] = useState(false);
  return jsx(FieldContainer, null, jsx(FieldLabel, null, field.label), onChange ? jsx(Fragment, null, jsx(TextInput, {
    autoFocus: autoFocus,
    disabled: onChange === undefined,
    onChange: event => {
      onChange(event.target.value);
    },
    onBlur: () => {
      setTouched(true);
    },
    placeholder: "0000-00-00T00:00:00.000Z",
    value: value
  }), (touched || forceValidation) && !isValid(value) && jsx("div", {
    css: {
      color: 'red'
    }
  }, "Timestamps must be in the form 0000-00-00T00:00:00.000Z")) : formatOutput(value));
};
const Cell = ({
  item,
  field,
  linkTo
}) => {
  let value = item[field.path];
  return linkTo ? jsx(CellLink, linkTo, formatOutput(value)) : jsx(CellContainer, null, formatOutput(value));
};
Cell.supportsLinkTo = true;
const CardValue = ({
  item,
  field
}) => {
  return jsx(FieldContainer, null, jsx(FieldLabel, null, field.label), item[field.path]);
};

function isValid(value) {
  if (value === '') return true;

  try {
    return parseISO(value).toISOString() === value;
  } catch (err) {
    return false;
  }
}

const controller = config => {
  return {
    path: config.path,
    label: config.label,
    graphqlSelection: config.path,
    defaultValue: '',
    deserialize: data => data[config.path] || '',
    serialize: value => ({
      [config.path]: value === '' ? null : value
    }),

    validate(value) {
      return isValid(value);
    }

  };
};

export { CardValue, Cell, Field, controller };
